pkg/chaoserr/common_errors.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2022 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package chaoserr
+
+import (
+	"github.com/pkg/errors"
+)
+
+type ErrNotFound struct {
+	name string
```

Comment:
> As [error handling rfc](https://github.com/chaos-mesh/rfcs/blob/main/text/2021-09-27-refine-error-handling.md) introduces, the name will never be extracted, so don't use a customized type for it.

https://github.com/Andrewmatilde/chaos-mesh/blob/tasks-manager/pkg/chaosdaemon/tasks/tasks_test.go#L110
As the test shows , developers actually need the name inside the ErrNotFound to indicate some info inside the task manager.
And I have leave a lot of detail message about what error will return , I think developers(maybe me) will must extract & handle most of them.
https://github.com/Andrewmatilde/chaos-mesh/blob/tasks-manager/pkg/chaosdaemon/tasks/task_manager.go#L127
https://github.com/Andrewmatilde/chaos-mesh/blob/tasks-manager/pkg/chaosdaemon/tasks/task_manager.go#L146
https://github.com/Andrewmatilde/chaos-mesh/blob/tasks-manager/pkg/chaosdaemon/tasks/task_manager.go#L176
https://github.com/Andrewmatilde/chaos-mesh/blob/tasks-manager/pkg/chaosdaemon/tasks/task_manager.go#L193

Reply:
When golang 1.18 comes , we can get full type support in it : 
```
type PID int
type ErrNotFound[T any] struct {
	value T
}

func (e *ErrNotFound[T]) Error() string {
	return fmt.Sprintf("%T not found: %v", e.value, e.value)
}

func NotFound[T any](value T) error {
	return &ErrNotFound[T]{value: value}
}
```

Reply:
How about making `name` exported? I think it might be used in other situations for error assertion not only in test codes.

Reply:
I agree with you.

Reply:
Even if you want to use a customized type, it would be better to use `const UidNotFound = ErrNotFound {Name: "UID"}` or `const UidNotFound = NotFound("UID")`, and assert with `errors.Is(x, chaoserr.UidNotFound)`, rather than use the function call `chaoserr.NotFound("UID")` everywhere. As you can see, the only argument in `chaoserr.NotFound` is never a real variable. It's always literal (and all possibilities can be enumerated in `chaoserr` package).

If you accept that a global constant variable is better than the function call statement, then there is little different between  `const UidNotFound = ErrNotFound {Name: "UID"}` and ` const UidNotFound = errors.New("uid not found")`. Consider this situation, the type `ErrNotFound` is obviously not needed anymore.

Reply:
As you have created `ErrCanNotAdd` and `ErrCanNotAssign`, you will not choose `ErrCanNot {Name: "Add"}`, right?

Reply:
I agree with that we need a global immutable error variable, but in golang ['constant can only be numbers, characters (runes), strings or booleans'](https://go.dev/doc/effective_go#constants). 
So `const UidNotFound = errors.New("uid not found")` is wrong code, I think we can write it in two way :

1. 
```
type ErrUidNotFound string
func (e *ErrUidNotFound ) Error() string
const ErrUidNotFound = ErrUidNotFound("uid not found")
```
or 
```
type ErrNotFound string
func (e *ErrNotFound ) Error() string
const ErrUidNotFound = ErrNotFound("uid")
```
2.
```
type ErrNotFound struct { private_arg string}
func (e *ErrNotFound ) Error() string
func NotFound(s string) ErrNotFound
var ErrUidNotFound = NotFound("uid")
```

I perfer the last one, but I export the `private_arg` , because I trust that if developers need it to be export here, it must have a suitable reason (some problem comes from Marshual but it not matter here). 
And I will make it unexported because now I find developer can just use interface Error to handle the situation.

> Considering this situation, the type ErrNotFound is obviously not needed anymore.

I think I need func NotFound just because I don't want to write not found every where without code auto-completion, and we can found a lot of constant string like "xxx not found","xxx is not found" or "not found xxx" here.

> As you have created `ErrCanNotAdd` and `ErrCanNotAssign`, you will not choose `ErrCanNot {Name: "Add"}`, right?

I think we should be careful about any abstract, but the question seems like why we do not give custom struct for every english word. 
I have carefully considered about it , so I only wrote two error in common_error:  'ERRNOTFOUND' and "ERRNOTIMPLEMENT" . I give  'ERRNOTFOUND'  because it is a commonly used error everywhere, 404(Not found) is the most famous use case, and for me I do need it because I do not want to write three "xxx not found".I give "ERRNOTIMPLEMENT" because I think you can just give something not implement directly to user or developer or yourself, so it is simple to use and will not add complexity to user.

Reply:
In conclusion, you will use `var errUidNotFound = NotFound("uid")`, right?

Another question is that, if every `NotFound("xxx")` has a corresponding variable, and are all asserted with `errors.Is`, the `NotFound` function can be implemented without the struct: `func NotFound(x string) error {return errors.Errorf("%s not found.", x);}`.

Reply:
> In conclusion, you will use var errUidNotFound = NotFound("uid")? I partially agree with it 👍

Yes.
> Another question is that, if every `NotFound("xxx")` has a corresponding variable, and are all asserted with `errors.Is`, the `NotFound` function can be implemented without the struct: `func NotFound(x string) error {return errors.Errorf("%s not found.", x);}`.

I aggree with you ,this will force developer use errors.Is to confirm every error NotFound. But for `ErrNotImplement` I will keep it ,because I think it is common to just return all the `ErrNotImplement`.